### DISCUSSION TITLE
Large Number of Small Files

### DISCUSSION THREAD

**Discussion by cddqssc** [pocketbase/pocketbase#5215] [Q&A]
Usage Scenario: PocketBase Extended with Go

Currently, I have approximately 3000 users, each of whom can upload up to 5000 files (with an average size of 500KB per binary file). Although the current user upload volume is low, with a total number of around 20,000 files, the number of uploads will increase over time. The maximum number of files could reach 3000 * 5000, which is 15 million files.

Additionally, each file will automatically create a corresponding .attrs file, resulting in a total of over 30 million files, not considering the potential growth in the number of users.

Given this enormous number of small files, will there be a significant performance degradation? Does anyone have related experience with this?

One solution I have considered is to create only one file per user. Each time a user uploads, the new data is appended to this file, with the offset and size recorded in the database. This way, the total number of files equals the number of users. Is this approach feasible?

Thanks for such an excellent project.

**Comment by ganigeorgiev** [comment]
I don't think anyone will be able to give you a definite answer as it would depends on too many things - type of disk, the used filesystem and its configuration options (eg. `ext4.large_disk`), OS inode limits and blocks size, etc.

In general, PocketBase doesn't really care how many files you have (in the database we store only the filenames). As long as you have the disk space and are allowed to upload that many files, I don't expect significant performance degradation in normal operations. In the most common case, when serving a file we perform a single db query to the related record by its id to produce its full path (aka. `collectionId/recordId/filename`).

Regarding the _"only one file per user"_ approach - I don't know how do you plan to make it work but I don't recommend it as I expect this to be slower compared to operating with individual files (for example deleting a "file" will require to modify that single large blob, which could lead to a lot of copying if the deleted part is not at the end; filelocks are also something that you have to keep in mind in case of multiple read/write access to that same file). There just too many edge cases with this approach and I don't see it benefits.

I remember you from https://github.com/pocketbase/pocketbase/discussions/4631 and I assume that you use Hetzner storage box for the uploaded files but I cannot find details about the limits of the service, nor information how it operates, so I'd advice you to contact Hetzner for more information.

**Reply by cddqssc** [reply]
Thank you for remembering me.

Due to business expansion, I am no longer using Hetzner but have instead set up a self-hosted physical server with the following specifications:

Bandwidth：300Mbps unlimited Network traffic
CPU: AMD 5700G
RAM: 2 * 32GB
NVME: 2 * 1TB, ZFS mirror, used for storing database-related files under the pb_data directory
HDD: 6 * 16TB, ZFS mirror, used for storing files under the pb_data/storage directory
After reviewing the relevant source code of Pocketbase, I agree with your opinion. Since the full file path (collectionId/recordId/filename) can be directly generated, file lookup should be quick, and performance will not significantly degrade with an increase in the number of files.

I also agree that the "only one file per user" method is not ideal because I did not consider the potential issues of large-scale duplication caused by user deletion operations and file locking-related problems.

Currently, I will continue to use Pocketbase's default file management method and will monitor whether there is a significant performance degradation as the number of files increases. If there is, I will update here.

Additionally, although I am using ZFS mirror, I will still back up the data to third-party cloud storage as a precaution. Currently, I am using rclone sync to periodically synchronize to Google Drive, but with the potential increase in file data in the future, I am unsure if Google Drive will allow the upload of a large number of small files. If not, I might create compressed files for pb_data/storage and then upload them. However, creating compressed files for a large number of small files is a cumbersome task that could take tens of hours and may result in corrupted compressed files.

For now, I am not overthinking it. I will continue to monitor the relationship between the number of files and performance, as well as the method for backing up a large number of files. If I find a better way, I will update it here.

Thank you again for such an outstanding project and your prompt response.

**Reply by thangngoc89** [reply]
@cddqssc Regarding backing up, I would suggest using a proper backup software like Borgbackup or restic with support for incremental backup. I've used borgbackup in the past but it's slow and doesn't support Windows out of the box. I'm using restic with a GUI called backrest.

**Reply by cddqssc** [reply]
> @cddqssc Regarding backing up, I would suggest using a proper backup software like Borgbackup or restic with support for incremental backup. I've used borgbackup in the past but it's slow and doesn't support Windows out of the box. I'm using restic with a GUI called backrest.

Looks cool, I'll check it out
